qiskit_ibm_runtime/noise_learner.py

```diff
@@ -0,0 +1,276 @@
+
+                * A :class:`Backend` if you are using job mode.
+                * A :class:`Session` if you are using session execution mode.
+                * A :class:`Batch` if you are using batch execution mode.
+
+                Refer to the `Qiskit Runtime documentation <https://docs.quantum.ibm.com/run>`_.
+                for more information about the ``Execution modes``.
+
+            backend: Backend to run the primitive. This can be a backend name or an :class:`IBMBackend`
+                instance. If a name is specified, the default account (e.g. ``QiskitRuntimeService()``)
+                is used.
+
+            session: Session in which to call the primitive.
+
+                If both ``session`` and ``backend`` are specified, ``session`` takes precedence.
+                If neither is specified, and the primitive is created inside a
+                :class:`qiskit_ibm_runtime.Session` context manager, then the session is used.
+                Otherwise if IBM Cloud channel is used, a default backend is selected.
+
+            options: :class:`NoiseLearnerOptions`. Alternatively, :class:`EstimatorOptions` can be
+                provided, in which case the estimator options get reformatted into noise learner options
+                and all the irrelevant fields are ignored.
+        """
+        self._mode: Optional[Union[Session, Batch]] = None
+        self._service: QiskitRuntimeService | QiskitRuntimeLocalService = None
+        self._backend: Optional[BackendV1 | BackendV2] = None
+
+        if isinstance(mode, (Session, Batch)):
+            self._mode = mode
+            self._service = self._mode.service
+            self._backend = self._mode._backend
+        elif isinstance(mode, IBMBackend):  # type: ignore[unreachable]
+            self._service = mode.service
+            self._backend = mode
+        elif isinstance(mode, (BackendV1, BackendV2)):
+            self._service = QiskitRuntimeLocalService()
+            self._backend = mode
+        elif isinstance(mode, str):
+            self._service = (
+                QiskitRuntimeService()
+                if QiskitRuntimeService.global_service is None
+                else QiskitRuntimeService.global_service
+            )
+            self._backend = self._service.backend(mode)
+        elif get_cm_session():
+            self._mode = get_cm_session()
+            self._service = self._mode.service
+            self._backend = self._service.backend(  # type: ignore
+                name=self._mode.backend(), instance=self._mode._instance
+            )
+        else:
+            raise ValueError("A backend or session must be specified.")
+
+        self._set_options(options)
+
+    @property
+    def options(self) -> NoiseLearnerOptions:
+        """The options in this noise learner."""
+        return self._options
+
+    def run(self, tasks: Iterable[Union[QuantumCircuit, EstimatorPubLike]]) -> RuntimeJobV2:
```

Comment:
I think we should consider this

```suggestion
    def run(self, circuits: Iterable[QuantumCircuit]) -> RuntimeJobV2:
```

with the recommendation

```python
learner.run(pub.circuit for pub in pubs)
```

My argument is that this will be a cleaner path to an interface that only accepts twirl blocks, assuming that's the goal.

Reply:
I am not 100% against this, because `pub.circuit for pub in pubs` is simple enough that it is not inconvenient.
But then I would also disallow passing estimator options (because it feels unnatural to allow est options but not pubs), and in that case we don't have a convenient way for users to do something like `learner_options.into_estimator_options`.

So if we make this change, IMO we should just change the narrative: The noise learner takes noise learner options and circuits, up to you to convert them

Reply:
Maybe a goal of the "experimental" stage of this feature can be to determine whether we want to have pub support. We allow pubs for now, and see how we feel about it as we continue to use it.

Reply:
Sounds like a plan

Reply:
If the goal is to see how useful pub support is, then it'd be nice to track the usage. For example, we can pass a new  `metadata` parameter to `noise-learner`, or lump it under `experimental`.